### PR TITLE
feat/add record batch reader registration

### PR DESCRIPTION
**Pull request by mesejo**

### CODE REVIEW

**Comment by dlovell** [comment]
@mesejo per our video call, we'd like to be able to optionally tee the record batches so we don't re-evaluate the remote table for each usage

File: python/letsql/tests/test_into_backend.py [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+
+@pytest.fixture(scope="session")
+def pg():
+    conn = ls.postgres.connect(
+        host="localhost",
+        port=5432,
+        user="postgres",
+        password="postgres",
+        database="ibis_testing",
+    )
```
Comment (dlovell) [inline python/letsql/tests/test_into_backend.py:37]: is this different than what `connect_env` would return?

**Comment by dlovell** [comment]
@mesejo I'm getting an error when I try to cache the trino-into-duckdb query

```
+ /nix/store/hfvg9bl7c7pyi4f59n9frmcxrj533hac-python3-3.10.14-env/bin/python -m pytest --import-mode=importlib -k _cached -- tests/test_into_backend.py
============================================================== test session starts ===============================================================
platform linux -- Python 3.10.14, pytest-8.0.2, pluggy-1.5.0
rootdir: /home/dan/repos/github/letsql
configfile: pyproject.toml
plugins: mock-3.14.0, snapshot-0.9.0, cov-5.0.0
collected 9 items / 8 deselected / 1 selected

tests/test_into_backend.py F                                                                                                               [100%]

==================================================================== FAILURES ====================================================================
_____________________________________________________ test_into_backend_duckdb_trino_cached ______________________________________________________

trino_table = r0 := DatabaseTable: orders
  orderkey      !int64
  custkey       !int64
  orderstatus   !string
  totalprice    !flo... orderpriority: r0.orderpriority
  clerk:         r0.clerk
  shippriority:  r0.shippriority
  comment:       r0.comment
tmp_path = PosixPath('/tmp/pytest-of-dan/pytest-149/test_into_backend_duckdb_trino0')

    def test_into_backend_duckdb_trino_cached(trino_table, tmp_path):
        db_con = ibis.duckdb.connect()
        expr = (
            trino_table
            .head(10_000)
            .pipe(into_backend, db_con)
            .pipe(make_merged)
            .cache(ParquetCacheStorage(path=tmp_path))
        )
>       df = ls.execute(expr)

tests/test_into_backend.py:224:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
expr/api.py:1610: in execute
    return con.execute(expr, **kwargs)
backends/let/__init__.py:228: in execute
    with self._get_backend_and_expr(expr) as resource:
/nix/store/a3dqdy161jnbfi66698yx7zfjd4cq4gw-python3-3.10.14/lib/python3.10/contextlib.py:135: in __enter__
    return next(self.gen)
backends/let/__init__.py:253: in _get_backend_and_expr
    expr = self._register_and_transform_cache_tables(expr)
backends/let/__init__.py:315: in _register_and_transform_cache_tables
    out = op.replace(fn)
/nix/store/hfvg9bl7c7pyi4f59n9frmcxrj533hac-python3-3.10.14-env/lib/python3.10/site-packages/ibis/common/graph.py:464: in replace
    results = self.map(replacer, filter=filter)
/nix/store/hfvg9bl7c7pyi4f59n9frmcxrj533hac-python3-3.10.14-env/lib/python3.10/site-packages/ibis/common/graph.py:265: in map
    results[node] = fn(node, results, **kwargs)
backends/let/__init__.py:307: in fn
    node = storage.set_default(uncached_to_expr, uncached)
common/caching.py:82: in set_default
    return self._put(key, default)
common/caching.py:141: in _put
    value.to_expr().to_parquet(loc)
/nix/store/hfvg9bl7c7pyi4f59n9frmcxrj533hac-python3-3.10.14-env/lib/python3.10/site-packages/ibis/expr/types/core.py:605: in to_parquet
    self._find_backend(use_default=True).to_parquet(self, path, **kwargs)
/nix/store/hfvg9bl7c7pyi4f59n9frmcxrj533hac-python3-3.10.14-env/lib/python3.10/site-packages/ibis/backends/duckdb/__init__.py:1452: in to_parquet
    query = self.compile(expr, params=params)
/nix/store/hfvg9bl7c7pyi4f59n9frmcxrj533hac-python3-3.10.14-env/lib/python3.10/site-packages/ibis/backends/sql/__init__.py:177: in compile
    query = self._to_sqlglot(expr, limit=limit, params=params, **kwargs)
/nix/store/hfvg9bl7c7pyi4f59n9frmcxrj533hac-python3-3.10.14-env/lib/python3.10/site-packages/ibis/backends/duckdb/__init__.py:109: in _to_sqlglot
    sql = super()._to_sqlglot(expr, limit=limit, params=params)
/nix/store/hfvg9bl7c7pyi4f59n9frmcxrj533hac-python3-3.10.14-env/lib/python3.10/site-packages/ibis/backends/sql/__init__.py:159: in _to_sqlglot
    sql = self.compiler.translate(table_expr.op(), params=params)
/nix/store/hfvg9bl7c7pyi4f59n9frmcxrj533hac-python3-3.10.14-env/lib/python3.10/site-packages/ibis/backends/sql/compiler.py:488: in translate
    results = op.map(fn)
/nix/store/hfvg9bl7c7pyi4f59n9frmcxrj533hac-python3-3.10.14-env/lib/python3.10/site-packages/ibis/common/graph.py:265: in map
    results[node] = fn(node, results, **kwargs)
/nix/store/hfvg9bl7c7pyi4f59n9frmcxrj533hac-python3-3.10.14-env/lib/python3.10/site-packages/ibis/backends/sql/compiler.py:468: in fn
    result = self.visit_node(node, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <ibis.backends.duckdb.compiler.DuckDBCompiler object at 0x7a2411b14700>
op = <letsql.expr.relations.MarkedRemoteTable object at 0x7a2411bc3460>
kwargs = {'name': 'ibis_remote-expr-placeholder_ecrqhjkuh5ef7o2ct6s44ohxnu_0', 'namespace': Namespace(catalog=None, database=No...te      date
  orderpriority  !string
  clerk          !string
  shippriority   !int32
  comment        !string
}, ...}
method = None

    def visit_node(self, op: ops.Node, **kwargs):
        if isinstance(op, ops.ScalarUDF):
            return self.visit_ScalarUDF(op, **kwargs)
        elif isinstance(op, ops.AggUDF):
            return self.visit_AggUDF(op, **kwargs)
        else:
            method = getattr(self, f"visit_{type(op).__name__}", None)
            if method is not None:
                return method(op, **kwargs)
            else:
>               raise com.OperationNotDefinedError(
                    f"No translation rule for {type(op).__name__}"
                )
E               ibis.common.exceptions.OperationNotDefinedError: No translation rule for MarkedRemoteTable

/nix/store/hfvg9bl7c7pyi4f59n9frmcxrj533hac-python3-3.10.14-env/lib/python3.10/site-packages/ibis/backends/sql/compiler.py:516: OperationNotDefinedError
============================================================ short test summary info =============================================================
FAILED tests/test_into_backend.py::test_into_backend_duckdb_trino_cached - ibis.common.exceptions.OperationNotDefinedError: No translation rule for MarkedRemoteTable
======================================================== 1 failed, 8 deselected in 0.67s =========================================================
```

File: python/letsql/tests/test_into_backend.py [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+
+    t = into_backend(pg.table("batting"), con, "ls_batting")
+
+    expr = (
+        t.join(t, "playerID")
+        .limit(15)
+        .cache(SourceStorage(source=con))
+        .pipe(into_backend, ddb_con)
+        .select(player_id="playerID", year_id="yearID_right")
+        .cache(ParquetCacheStorage(source=ddb_con))
```
Comment (dlovell) [inline python/letsql/tests/test_into_backend.py:137]: should this be using `path=tmp_path`?

**Comment by mesejo** [comment]
@dlovell The error is happening because the ibis duckdb backend doesn't know how to translate a MarkedRemoteTable, using ls.duckdb.connect should fix it

**Comment by mesejo** [comment]
> @mesejo per our video call, we'd like to be able to optionally tee the record batches so we don't re-evaluate the remote table for each usage

I've added teeing by default. Can we make the optionality available in another PR?

**Comment by mesejo** [comment]
> @dlovell The error happens because the ibis duckdb backend doesn't know how to translate a MarkedRemoteTable. Using `ls.duckdb.connect` should fix it.

I've added the test using `ls.duckdb.connect`

File: python/letsql/tests/test_into_backend.py [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+
+@pytest.fixture(scope="session")
+def pg():
+    conn = ls.postgres.connect(
+        host="localhost",
+        port=5432,
+        user="postgres",
+        password="postgres",
+        database="ibis_testing",
+    )
```
Comment (dlovell) [inline python/letsql/tests/test_into_backend.py:37]: should this be `connect_env`?

**Review by dlovell — APPROVED** [review]

**Pull request merged**